handshakes/src/types/covenant.rs

```diff
@@ -0,0 +1,300 @@
+        Self: std::marker::Sized,
+    {
+        let count = Self::read_compact_int(reader)?;
+
+        let mut items = vec![];
+        for _ in 0..count {
+            // TODO(mark): sane limit argument?
+            let item = CovenantItem::read_from(reader, 256)?;
+            items.push(item);
+        }
+
+        Ok(Self(items))
+    }
+
+    fn write_to<W>(&self, writer: &mut W) -> SerResult<usize>
+    where
+        W: Write,
+    {
+        let mut total: usize = 0;
+        total += Self::write_compact_int(writer, self.0.len() as u64)?;
+
+        for covenant_data in self.0.clone() {
+            total += covenant_data.write_to(writer)?;
+        }
+
+        Ok(total)
+    }
+}
+
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq)]
+/// A Covenant is a part of a Handshake UTXO that restricts how the UTXO may
+/// be spent over the period of multiple spends. It is able to maintain
+/// data over time.
+pub struct Covenant {
+    /// The CovenantType restricts spending to other CovenantTypes that have
+    /// a defined spending path. State machines can be constructed this way
+    /// that enable the auction system.
+    pub covenant_type: CovenantType,
+    /// The CovenantData represents Covenant specific data that is attached to
+    /// the UTXO. This data must be updated correctly as the UTXO is spent from
+    /// one CovenantType to the next CovenantType.
+    pub covenant_data: CovenantData,
+}
+
+impl Covenant {
+    /// Returns the null Covenant.
+    pub fn null() -> Self {
+        Self {
+            covenant_type: CovenantType::try_from("NONE").expect("Known covenant type"),
+            covenant_data: CovenantData::null(),
+        }
+    }
+}
+
+impl ByteFormat for Covenant {
+    type Error = SerError;
+
+    fn serialized_length(&self) -> usize {
+        let mut size: usize = 1; // covenant_type
+        size += self.covenant_data.serialized_length();
+        size
+    }
+
+    fn read_from<R>(reader: &mut R, _limit: usize) -> SerResult<Self>
+    where
+        R: Read,
+        Self: std::marker::Sized,
+    {
+        let mut buf = [0u8; 1];
+        reader.read_exact(&mut buf)?;
+        let covenant_type = u8::from_le_bytes(buf);
+        // TODO(mark): sane max?
+        let covenant_data = CovenantData::read_from(reader, 1024)?;
+
+        Ok(Self {
+            covenant_type: CovenantType::new(covenant_type),
+            covenant_data,
+        })
+    }
+
+    fn write_to<W>(&self, writer: &mut W) -> SerResult<usize>
+    where
+        W: Write,
+    {
+        let mut total: usize = 0;
+        let covenant_type: u8 = self.covenant_type.as_u8();
+        total += writer.write(&[covenant_type])?;
+        total += &self.covenant_data.write_to(writer)?;
+
+        Ok(total)
+    }
+}
+
+/// A Covenant Type defines the possible spend paths
+/// for a UTXO.
+#[derive(serde::Serialize, serde::Deserialize, Clone, Debug, Eq, PartialEq)]
+pub struct CovenantType(u8);
+
```

Comment:
The "safe" interface for creating a `CovenantType` is through `CovenantType::new`, there is the `try_from` interface to help make it so that consumers of the library don't need to memorize the types.

```rust
let covenant_type = CovenantType::try_from("BID").expect("");
```

What I'm not sure how to handle nicely is the fact that there are many `"UNKNOWN"` types and that an `"UNKNOWN"` could be defined in the future. The serialization code must use the `::new` interface to prevent the need of updating these `try_from` trait implementations.